### PR TITLE
refactor(x/ecocredit): add note and contract to origin tx

### PR DESCRIPTION
**Pull request by ryanchristo**
## Description

Ref: #1193 / Ref: #1206

This pull request adds note and contract to `OriginTx`. At the moment, nothing is done with contract on the server side, which is something we will store in state and handle in a followup. This pull request also refactors `OriginTx` validation to be a receiver method and moves to a separate file given `OriginTx` is no longer specific to `MsgMintBatchCredits`.

This pull request also removes note from `BatchOriginTx` (state object) because note was only intended as a field that would be passed to the events emitted by `MsgCreateBatch` and `MsgMintBatchCredits` (now included in `OriginTx`). Also `batch_denom` has been removed from `BatchOriginTx`, which serves no purpose.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/core/origin_tx.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+func (o *OriginTx) Validate() error {
+	if !reOriginTxId.MatchString(o.Id) {
+		return sdkerrors.ErrInvalidRequest.Wrap("origin_tx.id must be at most 128 characters long, valid characters: alpha-numberic, space, '-' or '_'")
+	}
+
+	if !reOriginTxSource.MatchString(o.Source) {
+		return sdkerrors.ErrInvalidRequest.Wrap("origin_tx.source must be at most 32 characters long, valid characters: alpha-numberic, space, '-' or '_'")
+	}
+
+	if len(o.Contract) > 0 && !isValidEthereumAddress(o.Contract) {
```
Comment (technicallyty) [inline x/ecocredit/core/origin_tx.go:21]: the `isValidEthereum` regexp catches the `>0` condition, so could simplify here

**Review by technicallyty — APPROVED** [review]
lgtm, 1 nit

File: x/ecocredit/core/origin_tx.go (hunk 1 above)
Comment (ryanchristo) [reply x/ecocredit/core/origin_tx.go:21]: So this is checking valid only if not empty. In some cases, we don't want contract to be required. For example, using mint batch credits when dynamically minting to a credit batch, which could be used as a manual bridge operation when bridging credits from a non-chain registry such as verra. https://github.com/regen-network/regen-ledger/pull/1206#discussion_r904122339

Comment (ryanchristo) [reply x/ecocredit/core/origin_tx.go:21]: We could make this required when a param is set to true (e.g. `contractRequired` or something) or if we make this always required we would only be supporting EVM-based chains wherever origin tx is being used.

Comment (technicallyty) [reply x/ecocredit/core/origin_tx.go:21]: ahhh right - i keep forgetting about the manual case 😖

**Review by clevinson — APPROVED** [review]
lgtm! Just one note that we should add credit class to `BatchOriginTx` table for security purposes, but happy to have that taken on in a follow-up.

File: proto/regen/ecocredit/v1/state.proto [hunk 2]
```diff
@@ -284,10 +284,4 @@ message BatchOriginTx {
   // source is the source chain or registry of the transaction originating
   // the mint process (e.g. polygon, ethereum, verra).
   string source = 2;
-
-  // note is a reference note for accounting that is passed to an event.
-  string note = 3;
-
-  // batch_denom is the denom of the credit batch.
-  string batch_denom = 4;
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply proto/regen/ecocredit/v1/state.proto:292]: Make sense. Would prefer to use `class_key` for efficient storage but same idea and I agree.

**Pull request merged**